core/lib/vm_utils/src/lib.rs

```diff
@@ -94,3 +43,36 @@ pub fn execute_tx<S: WriteStorage>(
     }
     Ok(())
 }
+
+pub async fn create_vm_for_l1_batch<H: HistoryMode>(
+    l1_batch_number: L1BatchNumber,
+    l2_chain_id: L2ChainId,
+    rt_handle: Handle,
+    mut connection: StorageProcessor<'_>,
+) -> anyhow::Result<VmAndStorage<H>> {
+    let prev_l1_batch_number = l1_batch_number - 1;
+    let (_, miniblock_number) = connection
+        .blocks_dal()
+        .get_miniblock_range_of_l1_batch(prev_l1_batch_number)
+        .await?
+        .with_context(|| {
+            format!(
+                "l1_batch_number {l1_batch_number:?} must have a previous miniblock to start from"
+            )
+        })?;
+
+    let mut pg_storage =
+        PostgresStorage::new(rt_handle.clone(), connection, miniblock_number, true);
+
+    let vm_env = VmEnvBuilder::new(l1_batch_number, u32::MAX, l2_chain_id)
+        .with_miniblock_number(miniblock_number)
+        .build(pg_storage.connection())
```

Comment:
You have the connection already, so you can just swap the lines so that you create `vm_env` first, and then create `pg_storage`. This way you don't need to expose `connection` method on the `PostgresStorage`.